doc/dev/error-categorization.md

```diff
@@ -0,0 +1,111 @@
+the reaction here would be to retry with a time based approach. On the other hand
+it does not make sense to differentiate between `No Write Permission` and `No Read Permission` as
+the application just knows that it simply cannot access the desired resource and tells
+the user to grant it. One could argue that an application could always react differently
+based on the error (eg. give yourself write or read permission) but such detailed granularity
+is rarely needed and would come with a lot of maintenance effort. In the previous versions of Elektra
+we were facing many duplicated errors because too many existed and developers simply cannot know each
+and every code.
+
+So if you feel the need for a new category you should first ask yourself the following question:
+> Can I react differently if I would know every error X in category Y?
+
+If not it will most likely fit into an existing category. Categories are leaf based, so you cannot
+put an error into a node such as `Permanent errors` (see below).
+
+## Error categorization Guideline
+
+Now we will investigate each category in more detail and when to put an error/warning in there.
+
+The current structure looks like this:
+- Permanent errors ("01000")
+  - Resource ("01100")
+  - Parsing ("01200")
+  - Installation ("01300")
+  - Logical ("01400")
+- Conflict ("02000")
+- Timeout ("03000")
+- Validation ("04000")
+  - Syntactic ("04100")
+  - Semantic ("04200")
+### Permanent errors ("01000")
+
+The main category `Permanent Errors` refer to such errors which cannot be fixed by retry 
+at all. `Permanent Errors` are subdivided into
+- Resource
+- Parsing
+- Installation
+- Logical
+#### Resource ("01100")
+
+`Resource Errors` are all kinds of errors which are either permission related, existence related
+or missing resources such as memory, RAM, etc.
+Examples are missing files, insufficient permissions or out of memory.
+This category forces users or applications to provide additional resources, create a missing file/ directory etc
+or provide the relevant permissions for Elektra to work.
+
+#### Parsing ("01200")
+
+`Parsing Errors` are reserved for parsers for xml/yaml/ini etc. and indicate errors which
+were investigated during parsing. Such errors show unexpected encounters like missing line endings,
+illegal characters or wrong encodings.
+Users or applications will have to sanitize the input and retry again.
+
+#### Installation ("01300")
+
+`Installation Errors` are those errors which are related to Elektra's installation such as
+wrong plugin names, missing backends, initialization errors, misconfiguration of Elektra etc.
+Users will have to reconfigure, reinstall, recompile (with other settings) Elektra in order to
+get rid of this error. Applications will most likely not be able to handle such errors by themselves.
+
+#### Logical ("01400")
+
+`Logical Errors` indicate a bug in Elektra such as internal errors, assertion failures or errors
```

Comment:
I think this should be two separate categories.

One for errors actually internal to Elektra like assertion failures (although assertion use `abort(3)` right now I think). These are the ones that come with a message to report the error to our bug tracker. 

And another one for logical interface errors. E.g. when an application passes a NULL pointer to `kdbGet`. In this case reporting the error to Elektra's bug tracker doesn't make sense, but an application might want to add a line telling user to notify the developer about the bug.